qualtran/resource_counting/_qubit_counting.py

```diff
@@ -0,0 +1,60 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Callable, Set, Union
+
+import networkx as nx
+import sympy
+
+from qualtran import Bloq, Connection, DanglingT
+from qualtran._infra.composite_bloq import _binst_to_cxns
+
+
+def _cbloq_max_width(
+    binst_graph: nx.DiGraph, _bloq_max_width: Callable[[Bloq], int] = lambda b: 0
+) -> Union[int, sympy.Expr]:
+    """Get the maximum width of a composite bloq.
+
+    Specifically, we treat each binst in series. The width at each inter-bloq time point
+    is the sum of the bitsizes of all the connections that are "in play". The width at each
+    during-a-binst time point is the sum of the binst width (which is provided by the
+    `_bloq_max_width` callable) and the bystander connections that are "in play". The max
+    width is the maximum over all the time points.
+    """
+    max_width: Union[int, sympy.Expr] = 0
+    series_i: int = 0
+    in_play: Set[Connection] = set()
+
+    for gen_i, binsts in enumerate(nx.topological_generations(binst_graph)):
+        for binst in binsts:
+            pred_cxns, succ_cxns = _binst_to_cxns(binst, binst_graph=binst_graph)
```

Comment:
I'd suggest to add an outer loop for analyzing each connected component independently, so we have (pseudocode) as follows

```
for cc_i, cc_binst_graph in enumerate_connected_components(binst_graph):
    for gen_i, binsts in enumerate(nx.topological_generations(cc_binst_graph)): 
        for binst in binsts:
            ....
```

This should result in output being 2 * n vs 3 * n for the two examples discussed in the issue

Reply:
I didn't understand that in your example. If you drew that circuit diagram with a random alloc and free happening alongside your calculation you'd say the width included that line. I guess if your bloq is composed of disconnected components, you could (theoretically) do each component sequentially so the max widths should be considered separately?